### PR TITLE
Supporting functionalities for direct migration

### PR DESCRIPTION
**Pull request by matt-tsai**
Main changes include: ReadWrite interface for transport TPMs, adding util to grab HandleValue.

### CODE REVIEW

File: direct/transport/tpm.go [hunk 1]
```diff
@@ -10,14 +10,32 @@ import (
...
 	Send(input []byte) ([]byte, error)
+}
+
+// TPMCloser represents a logical connection to a TPM and you can close it!!!.
+type TPMCloser interface {
+	TPM
 	Close() error
 }
 
+type tempTPM struct {
```
Comment (matt-tsai) [inline direct/transport/tpm.go:21]: Needs a better name...

File: direct/tpm2/tpm2.go [hunk 2]
```diff
@@ -534,7 +534,7 @@ type Sign struct {
 	// If keyHandle is not a restricted signing key, then this
 	// may be a NULL Ticket with tag =
 	// TPM_ST_CHECKHASH.
-	Validation tpmt.TKHashCheck
+	Validation tpmt.TKHashCheck
```
Comment (josephlr) [inline direct/tpm2/tpm2.go:537]: Why whitespace here?

File: direct/transport/tpm.go (hunk 1 above)
Comment (josephlr) [reply direct/transport/tpm.go:21]: `compatTPM`? legacyTPM?

**Review by josephlr — COMMENTED** [review]
One idea worth considering here is which direction we want to support conversion for the TPM types. We can support:
  1. Using the new TPM type (`transport.TPM`) with the old TPM interface (`io.ReadWriter`/`io.ReadWriteCloser`) and library.
  2. Using the old TPM type (`io.ReadWriter`) with the new TPM interface (`transport.TPM`) and library

This PR implements (2), but I'm wondering if we should implement (1) instead.

**Comment by matt-tsai** [comment]
> One idea worth considering here is which direction we want to support conversion for the TPM types. We can support:
> 
> 1. Using the new TPM type (`transport.TPM`) with the old TPM interface (`io.ReadWriter`/`io.ReadWriteCloser`) and library.
> 2. Using the old TPM type (`io.ReadWriter`) with the new TPM interface (`transport.TPM`) and library
> 
> This PR implements (2), but I'm wondering if we should implement (1) instead.

This PR implements 2 because the old tests still need RWC and the Direct Migration needs a RW in order to be made a transport tpm used for Execute. Correct me if I'm wrong but eventually we will only have 1 interface. 
I think the option (option 2 in this case) where we change RW into transport TPMs might be wiser as it would require less changes to direct and would also be backwards compatible.

File: direct/tpm2/tpm2.go (hunk 2 above)
Comment (matt-tsai) [reply direct/tpm2/tpm2.go:537]: Removed whitespace.

File: direct/transport/tpm.go (hunk 1 above)
Comment (alexmwu) [reply direct/transport/tpm.go:21]: seeing as we already have `LocalTPM`, we could also rename that to `LocalTPMCloser` and name this one `LocalTPM`

Comment (chrisfenner) [reply direct/transport/tpm.go:21]: I'll contribute to the bikeshed :)

If I understand @josephlr's out-of-band comments, we should return non-public types that happen to satisfy public interfaces. Therefore we need:

Our interfaces:
`TPM`
`TPMCloser`

Our private type names for each of these:
`localTPM` (implements both TPM and TPMCloser)
`wrappedTPM` (implements only TPM)
No wrappedTPMCloser because we shouldn't close a TPM whose readwriter we're borrowing

Don't need public type names like `WrappedTPM` or `LocalTPM` (so we can delete `LocalTPM` entirely). Just the public functions that return `TPM` or `TPMCloser` (`OpenSimulator`, `OpenTPM`, `WrapReadWriter`)

File: direct/transport/open_other.go [hunk 3]
```diff
@@ -22,7 +22,7 @@ func OpenTPM(path ...string) (TPM, error) {
 	if err != nil {
 		return nil, err
 	}
-	return &LocalTPM{
```
Comment (josephlr) [inline direct/transport/open_other.go:25]: Here (and in open_windows.go) we need the `OpenTPM` function to return `TPMCloser`.

File: direct/tpm2/tpm2.go [hunk 4]
```diff
@@ -532,8 +532,7 @@ type Sign struct {
 	InScheme tpmt.SigScheme
 	// proof that digest was created by the TPM
 	// If keyHandle is not a restricted signing key, then this
-	// may be a NULL Ticket with tag =
-	// TPM_ST_CHECKHASH.
+	// may be a NULL Ticket with tag = TPM_ST_CHECKHASH.
```
Comment (josephlr) [inline direct/tpm2/tpm2.go:535]: Is this change supposed to be in this PR?

File: direct/transport/tpm.go (hunk 1 above)
Comment (matt-tsai) [reply direct/transport/tpm.go:24]: I think Joe's comment makes the most sense to me too. :) Thank you for the suggestions

File: direct/transport/open_other.go (hunk 3 above)
Comment (matt-tsai) [reply direct/transport/open_other.go:25]: Fixed.

File: direct/transport/tpm.go [hunk 5]
```diff
@@ -10,20 +10,77 @@ import (
...
+}
+
+// The wrappedRWC represents a struct that wraps an io.ReadWriteCloser
+// to function like a transport.TPM with the close function.
+type wrappedRWC struct {
 	transport io.ReadWriteCloser
 }
 
+// The wrappedTPM represents a struct that wraps a transport.TPM 
+// to function like a io.ReadWriter
```
Comment (chrisfenner) [inline direct/transport/tpm.go:34]: nit: Maybe rephrase, like:

wrappedTPM represents a struct that wraps a transport.TPM's underlying transport to use with legacy code that expects an io.ReadWriter

File: direct/transport/tpm.go [hunk 6]
```diff
@@ -10,20 +10,77 @@ import (
...
+type wrappedTPM struct {
+	response []byte
+	tpm      TPM
+}
+
+// FromReadWritter takes in a io.ReadWriter TPM and returns a
+// transport.TPm wrapping the io.ReadWriter.
```
Comment (chrisfenner) [inline direct/transport/tpm.go:41]: nit: transport.TPM

File: direct/transport/tpm.go [hunk 7]
```diff
@@ -10,20 +10,77 @@ import (
...
 }
 
+// The wrappedTPM represents a struct that wraps a transport.TPM 
+// to function like a io.ReadWriter
+type wrappedTPM struct {
+	response []byte
+	tpm      TPM
+}
+
+// FromReadWritter takes in a io.ReadWriter TPM and returns a
```
Comment (chrisfenner) [inline direct/transport/tpm.go:40]: Nit: FromReadWriter

**Review by chrisfenner — APPROVED** [review]
Thanks for the change, looks great. Only teeny nits

File: direct/transport/tpm.go (hunk 5 above)
Comment (matt-tsai) [reply direct/transport/tpm.go:34]: 100% agreed, I was having trouble describing it myself, thank you!

File: direct/transport/tpm.go (hunk 6 above)
Comment (matt-tsai) [reply direct/transport/tpm.go:41]: Fixed

File: direct/transport/tpm.go (hunk 7 above)
Comment (matt-tsai) [reply direct/transport/tpm.go:40]: Fixed.

File: direct/transport/tpm.go [hunk 8]
```diff
@@ -10,20 +10,74 @@ import (
...
+
+// add a toreadwrite from transport
+func ToReadWriter(tpm TPM) io.ReadWriter {
+	return &wrappedTPM{tpm: tpm}
+}
+
+// Not sure what else to do here? 
+func (t *wrappedTPM) Read(p []byte) (n int, err error) {
+	return len(p), nil
+}
```
[the remark this answers is not in the document]
Comment (matt-tsai) [reply direct/transport/tpm.go]: Fixed.

File: direct/transport/tpm.go [hunk 9]
```diff
@@ -10,20 +10,77 @@ import (
...
+	return &wrappedTPM{tpm: tpm}
+}
+
+// Read copies t.response into the p buffer and return the appropriate length.
+func (t *wrappedTPM) Read(p []byte) (n int, err error) {
+	p = t.response
+	return len(p), nil
+}
+
+// Write writes len(p) bytes from p to the underlying data stream.
```
[the remark this answers is not in the document]
Comment (matt-tsai) [reply direct/transport/tpm.go:58]: Got it, updated this.

**Review by chrisfenner — APPROVED** [review]

File: direct/transport/tpm.go [hunk 10]
```diff
@@ -55,13 +56,7 @@ func (t *wrappedTPM) Read(p []byte) (n int, err error) {
...
 }
 
-// Write writes len(p) bytes from p to the underlying data stream.
-// It returns the number of bytes written from p (0 <= n <= len(p))
-// and any error encountered that caused the write to stop early.
-// Write must return a non-nil error if it returns n < len(p).
-// Write must not modify the slice data, even temporarily.
-//
-// Implementations must not retain p.
+// Write implements the io.ReadWriter interface
```
[the remark this answers is not in the document]
Comment (matt-tsai) [reply direct/transport/tpm.go:59]: Thanks for letting me know, I will keep this in mind going forward! Also it was been fixed.

File: direct/transport/tpm.go [hunk 11]
```diff
@@ -10,20 +10,72 @@ import (
...
+
+// ToReadWriter takes in a transport TPM and returns an
+// io.ReadWriter wrapping the transport TPM.
+func ToReadWriter(tpm TPM) io.ReadWriter {
+	return &wrappedTPM{tpm: tpm}
+}
+
+// Read copies t.response into the p buffer and return the appropriate length.
+func (t *wrappedTPM) Read(p []byte) (n int, err error) {
+	p = t.response
```
Comment (josephlr) [inline direct/transport/tpm.go:55]: So I don't think this works, using this implementation we won't actually write any bytes to p. This just overwrites p temporarily in the function, but doesn't actually do anything. Example:
https://go.dev/play/p/Lqdl8wT_1n1

I think we will need to either use a `*bytes.Buffer` as the internal response type (example: https://go.dev/play/p/JW4V94-0Blu), or have an implementation that looks like:
```go
func (t *wrappedTPM) Read(p []byte) (int, error) {
	n := copy(p, t.response)
	t.response = t.response[n:]
	if len(t.response) == 0 {
		return n, io.EOF
	}
	return n, nil
}
```
see this example: https://go.dev/play/p/Hppzq-rQETY
which uses [the `copy` builtin](https://pkg.go.dev/builtin@go1.18.3#copy)

Comment (matt-tsai) [reply direct/transport/tpm.go]: Fixed, should work as intended now.

**Comment by josephlr** [comment]
Nice work on this!!!
